File: epi_judge_java/epi/Knapsack.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.EpiUserType;
import epi.test_framework.GenericTest;

import java.util.List;

public class Knapsack {

    @EpiTest(testDataFile = "knapsack.tsv")
    public static int optimumSubjectToCapacity(List<Item> items, int capacity) {
        int[] dp = new int[capacity + 1];
        for (Item item : items) {
            for (int i = capacity; i >= item.weight; i--) {
                dp[i] = Math.max(dp[i], item.value + dp[i - item.weight]);
            }
        }
        int ans = 0;
        for (int d : dp) {
            ans = Math.max(ans, d);
        }
        return ans;
    }

    @EpiUserType(ctorParams = {Integer.class, Integer.class})
    public static class Item {
        public Integer weight;
        public Integer value;

        public Item(Integer weight, Integer value) {
            this.weight = weight;
            this.value = value;
        }
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "Knapsack.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/knapsack.cc
#include <vector>

#include "test_framework/generic_test.h"
#include "test_framework/serialization_traits.h"
using namespace std;

struct Item {
    int weight, value;
};

int OptimumSubjectToCapacity(const vector<Item> &items, int capacity) {
    vector<int> dp(capacity + 1, 0);
    for (auto &item : items) {
        for (int i = capacity; i >= item.weight; i--) {
            dp[i] = max(dp[i], dp[i - item.weight] + item.value);
        }
    }
    return *max_element(dp.begin(), dp.end());
}

namespace test_framework {
    template<>
    struct SerializationTrait<Item> : UserSerTrait<Item, int, int> {
    };
} // namespace test_framework

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"items", "capacity"};
    return GenericTestMain(args, "knapsack.cc", "knapsack.tsv",
                           &OptimumSubjectToCapacity, DefaultComparator{},
                           param_names);
}
